File: src/PerfTest/StructClass/StructClassTest.cs
using PerfTest.StructClass;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

namespace PerfTest.StructClass
{
    public class StructClassTest
    {
        const int CaseSize = 20 * 1024 * 1024;

        public static void DoTest()
        {
            var pt = new StructClassTest();
            //pt.EmptyMemory();
            //pt.ClassMemoryTest();
            //pt.StructMemoryTest();

            //pt.ClassTimeTest();
            //pt.StructTimeTest();

            //pt.ClassGenerateTimeTest();
            //pt.StructGenerateTimeTest();
        }

        public void EmptyMemory()
        {
            Console.WriteLine(nameof(EmptyMemory));
            Console.WriteLine($"Memory: {GetCurrentProcessMemoryMb()} MB.");
        }
        
        public void ClassMemoryTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(ClassMemoryTest));
            var buffer = Vector2C.CreateN(size).ToList();
            Console.WriteLine($"Memory: {GetCurrentProcessMemoryMb()} MB, size = {buffer.Count}.");
        }
        
        public void StructMemoryTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(StructMemoryTest));
            var buffer = Vector2S.CreateN(size).ToList();
            Console.WriteLine($"Memory: {GetCurrentProcessMemoryMb()} MB, size = {buffer.Count}.");
        }
        
        public void ClassTimeTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(ClassTimeTest));
            var buffer = Vector2C.CreateN(size).ToList();
            MeasureTime(() => buffer.Average(v => v.X + v.Y));
        }
        
        public void StructTimeTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(StructTimeTest));
            var buffer = Vector2S.CreateN(size).ToList();
            MeasureTime(() => buffer.Average(v => v.X + v.Y));
        }

        public void StructGenerateTimeTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(StructGenerateTimeTest));
            MeasureTime(() => Vector2S.CreateN(size).ToList());
        }

        public void ClassGenerateTimeTest(int size = CaseSize)
        {
            Console.WriteLine(nameof(ClassGenerateTimeTest));
            MeasureTime(() => Vector2C.CreateN(size).ToList());
        }

        private void MeasureTime<T>(Func<T> action)
        {
            var sw = new Stopwatch();
            sw.Start();
            var t = action();
            Console.WriteLine($"{sw.ElapsedMilliseconds}ms, result: {t}.");
        }

        private float GetCurrentProcessMemoryMb()
        {
            var ps = Process.GetCurrentProcess();
            return ps.WorkingSet64 / 1024.0f / 1024;
        }
    }
}

File: perf_test/StructClass/StructClassTest.cpp
#include "StructClassTest.h"
#include "Vector2C.h"
#include "Vector2S.h"
#include "..\Utils\Utils.h"

#include <iostream>
#include <algorithm>
#include <numeric>

using namespace std;

void StructClass::StructClassTest::DoTest()
{
	StructClassTest t;

	t.ShowMemory();

	t.ClassMemoryTest();
	t.StructMemoryTest();

	t.ClassTimeTest();
	t.StructTimeTest();

	t.GenerateClassTimeTest();
	t.GenerateStructTimeTest();
}

void StructClass::StructClassTest::ShowMemory()
{
	cout << "Memory: " << Utils::GetProcessMemorySizeMb() << " MB." << endl;
}

void StructClass::StructClassTest::ClassMemoryTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = Vector2C::CreateN(CaseSize);
	ShowMemory();
}

void StructClass::StructClassTest::StructMemoryTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = Vector2S::CreateN(CaseSize);
	ShowMemory();
}

void StructClass::StructClassTest::ClassTimeTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = Vector2C::CreateN(CaseSize);
	auto time = Utils::MeasureTime([&]() {
		auto avg = 1.0 * accumulate(begin(buffer), end(buffer), size_t{}, [](size_t a, shared_ptr<Vector2S> b) {
			return a + b->X + b->Y;
		}) / buffer.size();
		cout << "avg = " << avg << "." << endl;
	});
	cout << "time = " << time << "ms." << endl;
}

void StructClass::StructClassTest::StructTimeTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = Vector2S::CreateN(CaseSize);
	auto time = Utils::MeasureTime([&]() {
		auto avg = 1.0 * accumulate(begin(buffer), end(buffer), size_t{}, [](size_t a, Vector2S b) {
			return a + b.X + b.Y;
		}) / buffer.size();
		cout << "avg = " << avg << "." << endl;
	});
	cout << "time = " << time << "ms." << endl;
}

void StructClass::StructClassTest::GenerateClassTimeTest()
{
	cout << __FUNCTION__ << endl;
	auto time = Utils::MeasureTime([&]() { Vector2C::CreateN(CaseSize); });
	cout << "time = " << time << "ms." << endl;
}

void StructClass::StructClassTest::GenerateStructTimeTest()
{
	auto time = Utils::MeasureTime([&]() { Vector2S::CreateN(CaseSize); });
	cout << "time = " << time << "ms." << endl;
}
